packages/server-boilerplate/src/connections/EntityApi.ts

```diff
@@ -0,0 +1,366 @@
+      recurseFilter(value, filterArray, nestedKeys.concat(subKey)),
+    );
+    return filterArray;
+  }
+
+  if (Array.isArray(filter)) {
+    filterArray.push([nestedKeys, filter]);
+    return filterArray;
+  }
+
+  if (typeof filter === 'string') {
+    filterArray.push([nestedKeys, [filter]]);
+    return filterArray;
+  }
+
+  // Ignore all other values
+  return filterArray;
+};
+
+type Prefix<T, P extends string> = {
+  [field in keyof T & string as `${P}_${field}`]: T[field];
+};
+
+export class EntityApi implements MicroserviceApi {
+  public baseUrl = ENTITY_API_URL;
+
+  private readonly connection: ApiConnection;
+
+  constructor(connection: ApiConnection) {
+    this.connection = connection;
+  }
+
+  private stringifyFields(fields?: EntityApiQueryOptions['fields']) {
+    return fields ? fields.join(',') : undefined;
+  }
+
+  private stringifyFilter(filter?: EntityApiQueryOptions['filter']) {
+    return filter
+      ? recurseFilter(filter)
+          .map(
+            ([keys, values]) =>
+              `${keys.join(NESTED_FIELD_DELIMITER)}${FIELD_VALUE_DELIMITER}${values.join(
+                MULTIPLE_VALUES_DELIMITER,
+              )}`,
+          )
+          .join(CLAUSE_DELIMITER)
+      : undefined;
+  }
+
+  private stringifyQueryParameters(queryOptions?: EntityApiQueryOptions) {
+    if (!queryOptions) {
+      return undefined;
+    }
+
+    const { field, fields, filter } = queryOptions;
+    return { field, fields: this.stringifyFields(fields), filter: this.stringifyFilter(filter) };
+  }
+
+  private stringifyRelationshipsSubQueryParameters<Pref extends 'ancestor' | 'descendant'>(
+    queryOptions: RelationshipsSubQueryOptions | undefined,
+    prefix: Pref,
+  ) {
+    const stringifiedParameters = this.stringifyQueryParameters(queryOptions);
+    if (!stringifiedParameters) {
+      return undefined;
+    }
+
+    const { field, filter } = stringifiedParameters;
+    return {
+      [`${prefix}_field`]: field,
+      [`${prefix}_filter`]: filter,
+    } as Prefix<{ field?: string; filter?: string }, `${Pref}_`>;
+  }
+
+  public async getEntity(
+    hierarchyName: string,
+    entityCode: string,
+    queryOptions?: {
+      field: FlattableEntityFieldName;
+      filter?: EntityFilterObject;
+    },
+  ): Promise<FlattenedEntity>;
+  public async getEntity<T extends ExtendedEntityFieldName[]>(
+    hierarchyName: string,
+    entityCode: string,
+    queryOptions?: {
+      fields: T;
+      filter?: EntityFilterObject;
+    },
+  ): Promise<EntityApiResponse<T>>;
+  public async getEntity<T extends ExtendedEntityFieldName[]>(
+    hierarchyName: string,
+    entityCode: string,
+    queryOptions?: {
+      field?: FlattableEntityFieldName;
+      fields?: T;
+      filter?: EntityFilterObject;
+    },
+  ) {
+    return this.connection.get(`hierarchy/${hierarchyName}/${entityCode}`, {
+      ...this.stringifyQueryParameters(queryOptions),
+    });
+  }
```

Comment:
Typescript's horrendous method overriding syntax rears its ugly head again 😱 

However, I still feel it's worth it? Polymorphism is a hell of a drug and I don't wanna give it up 😝